addons/misra.py

```diff
@@ -256,6 +256,23 @@ def getForLoopExpressions(forToken):
             lpar.astOperand2.astOperand2.astOperand2]
 
 
+def getForLoopCounterVariables(forToken):
+    """
+    Return set of Variable objects used as for loop counters.
+    """
+    if not forToken or forToken.str != 'for':
+        return None
+    tn = forToken.next
+    if not tn or tn.str != '(':
+        return None
+    counterVars = set()
+    while tn.next and tn.next.str not in (';', ')'):
+        if tn.variable:
+            counterVars.add(tn.variable)
```

Comment:
wouldn't this mean that in such code `y` will be added as a for loop counter variable?

    for (x = y; x < 100; x++)

Reply:
I thought about that too. It is not very "precise". But I don't know if some clear definition of counter variable exists.
What is with such code:
```c
int x = 0;
for ( ; x < 5; ++x) {}
```
Maybe we could agree on a definition like this:
Variables that are assigned in the first for clause or changed in the third clause.?

Reply:
> wouldn't this mean that in such code `y` will be added as a for loop counter variable?
> 
> ```
> for (x = y; x < 100; x++)
> ```

No, here's a test: https://github.com/danmar/cppcheck/pull/2409/commits/1a41e6a2b950e1b37b6e6de06e99f30f5c489ec6.

Reply:
> I thought about that too. It is not very "precise". But I don't know if some clear definition of counter variable exists.
> What is with such code:
> 
> ```c
> int x = 0;
> for ( ; x < 5; ++x) {}
> ```
> 
> Maybe we could agree on a definition like this:
> Variables that are assigned in the first for clause or changed in the third clause.?

As I understand, 14.2 says about modification of variable declared in for scope, but I'm not sure. I add this [commentary](https://github.com/danmar/cppcheck/pull/2409/files#diff-1bfaaa2a6da0266311b76e7291e15236R260) to make it cleaner.

Reply:
> No, here's a test: 1a41e6a.

That feels good. But I don't understand why that works. Doesn't `getForLoopCounterVariables` get all variables before the first `; `?

Reply:
@danmar I see, the last variable was never checked. Fixed now, thanks.